modules/unsupported/jdbc-ng/jdbc-hana/src/main/java/org/geotools/data/hana/HanaFilterToSQL.java

```diff
@@ -0,0 +1,553 @@
+            DistanceBufferOperator filter,
+            PropertyName property,
+            Literal geometry,
+            boolean swapped,
+            Object extraData) {
+        if (!(filter instanceof DWithin) && !(filter instanceof Beyond)) {
+            throw new IllegalArgumentException("Unsupported filter type " + filter.getClass());
+        }
+        try {
+            property.accept(this, extraData);
+            out.write(".ST_WithinDistance(");
+            geometry.accept(this, extraData);
+            out.write(", ");
+            String unit = filter.getDistanceUnits();
+            if (UNITS_MAP.containsKey(unit)) {
+                double distInMeters = filter.getDistance() * UNITS_MAP.get(unit);
+                out.write(Double.toString(distInMeters));
+                out.write(", 'meter'");
+            } else {
+                out.write(Double.toString(filter.getDistance()));
+                out.write(", '");
+                out.write(filter.getDistanceUnits());
+                out.write("'");
+            }
+            out.write(")");
+            if ((filter instanceof DWithin) != swapped) {
+                out.write(" = 1");
+            } else {
+                out.write(" = 0");
+            }
+            return extraData;
+        } catch (IOException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    private Object visitBBOXSpatialOperator(
+            BBOX filter, PropertyName property, Literal geometry, Object extraData) {
+        try {
+            property.accept(this, extraData);
+            out.write(".ST_IntersectsRectPlanar(");
+            BoundingBox bbox = clamp(filter.getBounds());
+            Coordinate ll = new Coordinate(bbox.getMinX(), bbox.getMinY());
+            Coordinate ur = new Coordinate(bbox.getMaxX(), bbox.getMaxY());
+            GeometryFactory factory = new GeometryFactory();
+            visitGeometry(factory.createPoint(ll));
+            out.write(", ");
+            visitGeometry(factory.createPoint(ur));
+            out.write(") = 1");
+            if (filter instanceof BBOX3D) {
+                BBOX3D filter3d = (BBOX3D) filter;
+                BoundingBox3D bbox3d = filter3d.getBounds();
+                double minz = bbox3d.getMinZ();
+                double maxz = bbox3d.getMaxZ();
+                out.write(" AND ");
+                property.accept(this, extraData);
+                out.write(".ST_ZMin() <= ");
+                out.write(Double.toString(maxz));
+                out.write(" AND ");
+                property.accept(this, extraData);
+                out.write(".ST_ZMax() >= ");
+                out.write(Double.toString(minz));
+            }
+        } catch (IOException e) {
+            throw new RuntimeException(e);
+        }
+        return extraData;
+    }
+
+    private BoundingBox clamp(BoundingBox bounds) {
+        // In geographic CRS', HANA will reject any points outside the "normalized" range, which is
```

Comment:
This really comes from practical use of WMS/WFS, clients will normally do out of range requests, and admins will often record as WGS84 stuff that's really projected, and expect the result to work anyways.